src/python/pants/core/util_rules/external_tool.py

```diff
@@ -0,0 +1,248 @@
+            default=cls.default_known_versions,
+            advanced=True,
+            help=f"Known versions to verify downloads against. Each element is a "
+            f"pipe-separated string of version|platform|sha256|length, where `version` is the "
+            f"version string, `platform` is one of [{','.join(Platform.__members__.keys())}], "
+            f"`sha256` is the 64-character hex representation of the expected sha256 digest of the "
+            f"download file, as emitted by `shasum -a 256`, and `length` is the expected length of "
+            f"the download file in bytes. E.g., '3.1.2|darwin|"
+            f"6d0f18cd84b918c7b3edd0203e75569e0c7caecb1367bbbe409b44e28514f5be|42813'. "
+            f"Values are space-stripped, so pipes can be indented for readability if necessary."
+            f"You can compute the length and sha256 easily with:  "
+            f"curl -L $URL | tee >(wc -c) >(shasum -a 256) >/dev/null",
+        )
+        # For backwards compatibility with BinaryToolBase.  Subclasses of BinaryToolBase can
+        # instead subclass this class, and need only switch their --version-digest-mapping to
+        # --known-versions to get rid of the deprecation warning.
+        # Note: After removing this, also remove get_default_versions_and_digests() above.
+        register(
+            "--version-digest-mapping",
+            removal_version="1.30.0.dev0",
+            removal_hint="Use --known-versions instead.",
+            type=dict,
+            default=cls.get_default_versions_and_digests(),
+            fingerprint=True,
+            help="A dict mapping <platform constraint> -> (<version>, <fingerprint>, <size_bytes>)."
+            f'A "platform constraint" is any of {[c.value for c in PlatformConstraint]}, and '
+            "is the platform to fetch the tool for. A platform-independent tool should "
+            f"use {PlatformConstraint.none.value}, while a platform-dependent tool should specify "
+            'all environments it needs to be used for. The "fingerprint" and "size_bytes" '
+            "arguments are the result printed when running `sha256sum` and `wc -c` on "
+            "the downloaded file, respectively.",
+        )
+
+    @classmethod
+    @abstractmethod
+    def generate_url(cls, plat: Platform, version: str) -> str:
+        """Returns the URL for the given version of the tool, runnable on the given os+arch.
+
+        os and arch default to those of the current system.
+
+        Implementations should raise ExternalToolError if they cannot resolve the arguments
+        to a URL.  The raised exception need not have a message - a sensible one will be generated.
+        """
+
+    @classmethod
+    def generate_exe(cls, plat: Platform, version: str) -> str:
+        """Returns the archive path to the given version of the tool.
+
+        If the tool is downloaded directly, not in an archive, this can be left unimplemented.
+        """
+        return ""
```

Comment:
Consider `raise NotImplementedError`?

Reply:
This is actually a useful sentinel value detected by the calling code. It could catch an exception, but this isn't an "exceptional" situation really...